### PR TITLE
Package fingerprint storage

### PR DESCRIPTION
**Pull request by yim-lee**
Motivation:
Provide trust-on-first-use (TOFU) security model for package dependencies. SwiftPM will record locally the "fingerprint" of a package version when it is first downloaded and ensure the fingerprint remain the same in subsequent downloads.

For package downloaded from registry, the fingerprint is the source archive checksum. For source control, it is the git revision.

This PR only adds storage APIs for writing and reading package fingerprints and an implementation that uses local file system. The integration of fingerprint storage into the download workflow(s), i.e. TOFU implementation, will come in a separate PR.

Modifications:
- Add `PackageFingerprint` module.
- Add `PackageFingerprintStorage` protocol that defines APIs for reading and writing fingerprints.
- Add `FilePackageFingerprintStorage` which is an implementation based on file system.

### CODE REVIEW

**Comment by yim-lee** [comment]
@swift-ci please smoke test

File: Tests/PackageFingerprintTests/FilePackageFingerprintStorageTests.swift [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+                                                callback: callback) }
+        try tsc_await { callback in storage.put(package: package, version: Version("1.0.0"),
+                                                fingerprint: .init(origin: .sourceControl(sourceControlURL), value: "gitHash-1.0.0"),
+                                                callback: callback) }
+        try tsc_await { callback in storage.put(package: package, version: Version("1.1.0"),
+                                                fingerprint: .init(origin: .registry(registryURL), value: "checksum-1.1.0"),
+                                                callback: callback) }
+        // Fingerprint for another package
+        let otherPackage = PackageIdentity.plain("other.LinkedList")
+        try tsc_await { callback in storage.put(package: otherPackage, version: Version("1.0.0"),
```
Comment (tomerd) [inline Tests/PackageFingerprintTests/FilePackageFingerprintStorageTests.swift:39]: consider creating private extension with blocking versions to simply these call sites

**Review by tomerd — APPROVED** [review]

**Review by neonichu — APPROVED** [review]

File: Sources/PackageFingerprint/FilePackageFingerprintStorage.swift [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+public struct FilePackageFingerprintStorage: PackageFingerprintStorage {
+    let fileSystem: FileSystem
+    let directory: AbsolutePath
+
+    private let encoder: JSONEncoder
+    private let decoder: JSONDecoder
+
+    init(customFileSystem: FileSystem? = nil, customDirectory: AbsolutePath? = nil) {
+        self.fileSystem = customFileSystem ?? localFileSystem
+        self.directory = customDirectory ?? self.fileSystem.dotSwiftPM.appending(component: "fingerprints")
```
Comment (tomerd) [inline Sources/PackageFingerprint/FilePackageFingerprintStorage.swift:27]: the default location can also be encoded into Workspace.Location / DefaultLocation which handles similar set ups

Comment (yim-lee) [reply Sources/PackageFingerprint/FilePackageFingerprintStorage.swift]: Making both required so no need to have defaults here.

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Review by tomerd — APPROVED** [review]

**Comment by compnerd** [comment]
@yim-lee @tomerd - this regressed the Windows builds.  Please always make sure you coordinate any new dependencies with the Windows installers.

**Comment by yim-lee** [comment]
Sorry @compnerd, is there anything I need to do to help fix the Windows builds?

**Comment by compnerd** [comment]
Thanks @yim-lee.  I'll rest to see if I can put up a change later (tomorrow likely).  Is there any new target other than PackageFingerprint or any new dependencies?

**Comment by yim-lee** [comment]
@compnerd There are: 
- New module `PackageFingerprint` (this PR) 
- New dependencies
  - https://github.com/apple/swift-system (https://github.com/apple/swift-package-manager/pull/3881)
  - https://github.com/apple/swift-collections (pending - https://github.com/apple/swift-package-manager/pull/3681)

Does https://github.com/apple/swift-package-manager/pull/3968 address issue caused by `PackageFingerprint`?

**Comment by compnerd** [comment]
The problem is due to packaging, so there needs to be an associated change in apple/swift-installer-scripts to package up the additional DLL.  Assuming that it is just SystemPackage and OrderedCollections that are used, I think that its just the single missing dll.

**Comment by compnerd** [comment]
https://github.com/apple/swift-installer-scripts/pull/69 _should_ be what is needed.

**Pull request merged**